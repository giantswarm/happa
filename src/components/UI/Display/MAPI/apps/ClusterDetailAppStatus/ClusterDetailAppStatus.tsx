import { Box, Text } from 'grommet';
import * as applicationv1alpha1 from 'model/services/mapi/applicationv1alpha1';
import React from 'react';
import styled from 'styled-components';

const Title = styled(Text)`
  &::first-letter {
    text-transform: uppercase;
  }
`;

function getStatusColor(status: string) {
  switch (status) {
    case applicationv1alpha1.statusUnknown:
    case applicationv1alpha1.statusUninstalled:
      return { background: 'status-unknown', text: 'text-accent' };

    case applicationv1alpha1.statusDeployed:
      return { background: 'status-ok', text: 'text-strong' };

    case applicationv1alpha1.statusSuperseded:
    case applicationv1alpha1.statusUninstalling:
    case applicationv1alpha1.statusPendingInstall:
    case applicationv1alpha1.statusPendingUpgrade:
    case applicationv1alpha1.statusPendingRollback:
      return { background: 'status-warning', text: 'text-accent' };

    default:
      return { background: 'status-danger', text: 'text-strong' };
  }
}

function formatStatus(status: string) {
  return status.replace(/-/g, ' ');
}

interface IClusterDetailAppStatusProps
  extends React.ComponentPropsWithoutRef<typeof Box> {
  status: string;
}

const ClusterDetailAppStatus: React.FC<IClusterDetailAppStatusProps> = ({
  status,
}) => {
  const color = getStatusColor(status);
  const title = formatStatus(status);

  return (
    <Box
      background={color.background}
      round='xxsmall'
      pad={{ vertical: 'xxsmall', horizontal: 'small' }}
    >
      <Title color={color.text} aria-label={`App status: ${title}`}>
        {title}
      </Title>
    </Box>
  );
};

export default ClusterDetailAppStatus;
