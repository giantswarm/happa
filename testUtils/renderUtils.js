import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
