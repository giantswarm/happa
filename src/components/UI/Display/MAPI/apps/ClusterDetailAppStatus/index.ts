import ClusterDetailAppStatus from './ClusterDetailAppStatus';

export default ClusterDetailAppStatus;
