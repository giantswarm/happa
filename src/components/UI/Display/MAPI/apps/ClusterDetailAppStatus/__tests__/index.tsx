import { renderWithTheme } from 'test/renderUtils';

import ClusterDetailAppStatus from '..';

describe('ClusterDetailAppStatus', () => {
  it('renders an app status', () => {
    const { container } = renderWithTheme(ClusterDetailAppStatus, {
      status: 'some-status',
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
