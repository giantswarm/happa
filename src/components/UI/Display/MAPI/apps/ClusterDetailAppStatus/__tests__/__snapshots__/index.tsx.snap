// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClusterDetailAppStatus renders an app status 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 kQNwCm"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 fmpnGZ"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 dEwcnG"
    >
      <span
        aria-label="App status: some status"
        class="StyledText-sc-1sadyjn-0 gqqUFB ClusterDetailAppStatus__Title-sc-kfzx37-0 icerTK"
      >
        some status
      </span>
    </div>
  </div>
</div>
`;
